add encrypted storage feature and deprecate older features

bcressey opened the pull request:
**Issue number:**
Related: https://github.com/bottlerocket-os/bottlerocket/issues/4680

**Description of changes:**
Add an image feature for encrypted storage.

My plan is for this to cover both `/local` (using dm-crypt) and `/.bottlerocket/datastore` (using fscrypt). I'll post the related core kit changes soon.

Deprecate two image features that are no longer useful.

`aws-ecs-1` was the last variant where GRUB's `$private` variable wasn't set in the earliest version, and also the last one to use `wicked` instead of `systemd-networkd`. Since `aws-ecs-1` is gone, and `wicked` is gone, there's no need to keep these features around.


**Testing done:**
Confirmed that I could add the encrypted storage feature to a variant and trigger the install of supporting software.

Verified that software that requires the networkd image feature continues to install.


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
twoliter: add encrypted-storage image feature

Introduce a new, experimental image feature to encrypt local storage.
This is only compatible with platforms that provide a TPM2 device, so
that the required keys can be protected.

To avoid trivial offline attacks, where encrypted storage could be
replaced by unencrypted data in the expected format, any policy for
encryption needs to be defined at the variant level and then enforced
unconditionally by the OS.

As with most other image features, enabling this flag for an existing
variant is backwards-incompatible, and migrations into or out from an
encrypted state are not supported.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
buildsys: support deprecating image features

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
twoliter: deprecate grub-set-private-var image feature

The GRUB change to set the $private variable, so that grub.cfg can
point to bootconfig data, has been in place since early 2022.

Mark the feature as deprecated to signal downstream builds that the
image feature can safely be removed from variant definitions.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
twoliter: deprecate systemd-networkd image feature

wicked is no longer packaged for Bottlerocket, and systemd-networkd
is the only supported network management daemon.

Mark the feature as deprecated to signal downstream builds that the
image feature can safely be removed from variant definitions.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

cbgbt commented on tools/buildsys/src/manifest.rs:
```diff
@@ -821,6 +826,8 @@ pub enum ImageFeature {
 
 const EXPERIMENTAL_IMAGE_FEATURES: [&ImageFeature; 0] = [];
 
+const DEPRECATED_IMAGE_FEATURES: [&ImageFeature; 2] = [];
```
You can avoid having to specify the size of the array by making this a const reference:

```suggestion
const DEPRECATED_IMAGE_FEATURES: &[&ImageFeature] = [];
```

cbgbt reviewed (approved):
Only a minor style nit. iirc this code is consistent with what exists today.

lgtm!

Commit:
twoliter: record in-place updates state in image-features.env

This allows programs to query whether in-place updates are enabled
without relying on other indicators that might change over time, such
as whether certain partitions are present.

In the specific case of encrypted storage, when in-place updates are
disabled, we can use more of the PCRs to seal encryption keys, since
the kernel and userspace should never change. Since this decision has
to be made before the API server's datastore can be created, we can't
rely on an API to discover the feature state.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

bcressey commented:
⬆️ force push fixes a tiny whitespace nit
